at_cmd_parser: Add support for TEST and READ command parsing

rlubos opened the pull request:
Several changes included in this PR:

* I've run clang-format over at_cmd_parser sources to improve coding standard compatibility,
* Small bugfix in one of the at_cmd_parser test cases,
* Add support for READ and TEST command parsing. So far, only SET was supported.

To differentiate between commands, I've introduced new AT param types: `AT_PARAM_TYPE_CMD_*`. Depending on the command type, the first parameter parsed, holding the command string, will be one of these types. This allows the application to easily determine which type of command it deals with after parsing, w/o a need to manually check for special characters (`?` or `=?`).

This change will be used by the `serial_lte_modem` sample.

NordicBuilder commented:
**All checks passed.**

## checkpatch (informational only, not a failure)

```
No additional types will be considered - file 'scripts/checkpatch/typedefsfile': No such file or directory
-:485: WARNING:LONG_LINE: line over 80 characters
#485: FILE: tests/lib/at_cmd_parser/at_cmd_parser/src/main.c:546:
+	zassert_true(at_params_type_get(&test_list2, 0) == AT_PARAM_TYPE_CMD_SET,

-:504: WARNING:LONG_LINE: line over 80 characters
#504: FILE: tests/lib/at_cmd_parser/at_cmd_parser/src/main.c:565:
+	zassert_true(at_params_type_get(&test_list2, 0) == AT_PARAM_TYPE_CMD_SET,

-:515: WARNING:LONG_LINE: line over 80 characters
#515: FILE: tests/lib/at_cmd_parser/at_cmd_parser/src/main.c:594:
+	zassert_true(at_params_type_get(&test_list2, 0) == AT_PARAM_TYPE_CMD_SET,

-:550: WARNING:LONG_LINE: line over 80 characters
#550: FILE: tests/lib/at_cmd_parser/at_cmd_parser/src/main.c:648:
+	zassert_true(at_params_type_get(&test_list2, 0) == AT_PARAM_TYPE_CMD_READ,

-:586: WARNING:LONG_LINE: line over 80 characters
#586: FILE: tests/lib/at_cmd_parser/at_cmd_parser/src/main.c:684:
+	zassert_true(at_params_type_get(&test_list2, 0) == AT_PARAM_TYPE_CMD_TEST,

- total: 0 errors, 5 warnings, 558 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.

NOTE: Ignored message types: AVOID_EXTERNS BRACES CONFIG_EXPERIMENTAL CONST_STRUCT DATE_TIME FILE_PATH_CHANGES MINMAX MULTISTATEMENT_MACRO_USE_DO_WHILE NETWORKING_BLOCK_COMMENT_STYLE PRINTK_WITHOUT_KERN_LEVEL SPLIT_STRING VOLATILE

NOTE: If any of the errors are false positives, please report
      them to the maintainers.

```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

metc commented on lib/at_cmd_parser/at_params.c:
```diff
@@ -29,6 +29,9 @@ static void at_param_clear(struct at_param *param)
 	__ASSERT(param != NULL, "Parameter cannot be NULL.");
 
 	if ((param->type == AT_PARAM_TYPE_STRING) ||
+	    (param->type == AT_PARAM_TYPE_CMD_SET) ||
```
For set, read and test, the value is a normal String value. We already offer this type of value. Why adding a new type?
Instead, could you use another field/enum to indicates if it is a set, read or test string?

rlubos replied:
I need to carry the information about the type of command being processed - in this approach I confined this information in the param type of the first parameter.

But indeed, giving it a second thought, it might actually be an overuse to include this information in the parameter type, and indeed a better option would be to introduce a separate parameter for this case.

So what now I am thinking of is to provide an enumeration type:
```
enum at_cmd_type {
    AT_CMD_TYPE_SET_COMMAND,
    AT_CMD_TYPE_READ_COMMAND,
    AT_CMD_TYPE_TEST_COMMAND,
    AT_CMD_TYPE_NOTIFICATION,
}
```
and then, within the `struct at_param_list` add a new field: `enum at_cmd_type cmd_type;`.

This approach should be way less intrusive and actually neater IMO. What do you think?

metc replied:
I like this approach better. I do not think we need a new type.
Do we really need to store the command type in the parameter? All we need to know is that it is a String.
Can we have set and read parameters in the same list for instance? I think this information should only be available for the parser and is not related to parameters.
I looked at your updated test case, but I am not sure what the requirements really are.

rlubos replied:
Ok, let me explain.

We have a project called `serial_lte_modem` which extends the list of available AT commands with a set of additional commands (let's call them `custom` as `proprietary` is already used by Nordic AT commands specification). This project uses `at_cmd_parser` module to extract the actual command parameters.

So far, as we have no support for READ and TEST commands, the sample analyzed the command string manually to determine the type of command it deals with and then used the `at_cmd_parser` to extract the command parameters only for the SET command (READ/TEST carry no parameters other than the command itself). This approach is cumbersome though, as it expects the `at_cmd_parser` user to understand the AT command syntax and actually requires to parse part of the command twice.

This PR intends to facilitate the process so that users of `at_cmd_parser` module can simply parse the command string and get all the information needed. Let's consider the custom `at#xsocket` command, it can have one of three formats (which complies to the AT command specification):
1) SET: `at#xsocket=1,1`
2) READ: `at#xsocket?`
3) TEST: `at#xsocket=?`

Running the `at_cmd_parser` over these command strings will give the following results if we ignore the command type (I've specified the param type in the parentheses):
1) `"at#xsocket"(string) 1(short) 1(short)`
2) `"at#xsocket"(string)`
3) `"at#xsocket"(string)`

As you can see, based only on the existing parameter types I can tell no difference between 2) and 3). If the actual command accepted SET command w/o parameters, 1) would be problematic as well. So we need extra information, informing the user what type of command was just parsed.

The original idea was to store the command type within the first parameter type, holding the actual command string, but we've given up on this. The current idea is to store the command type within the `struct at_param_list` as a separate field. There's no use providing this information at per-parameter basis, as it will not differ across the parameters (it's the command feature, not a paramter feature). The user could then easily access this information with a getter function, after the command was parsed, w/o diving into command internals.

metc replied:
Thank you for the information.
We agree that adding a type for each parameter is not what we need.

I do not think the AT params list needs to know if the AT string is a set, test or read command. This information is known by the parser, so it should be stored in the parser, not in the parameter list.
What about adding an output parameter in the parser API?

```
int at_parser_params_from_str(const char *at_params_str, char **next_param_str,
    struct at_param_list *const list,
    const enum *const at_cmd_type);
```

rlubos replied:
If you prefer to modify the API and provide command type as a separate output parameter, sure I can do it. Just note this will make the PR much larger as I will need to align all of the existing API calls.

I'm still not convinced though it's the right way to go. `struct at_param_list` is nothing more than a dissected AT command/notification and I see no harm in extending it with additional parser products. But I will not push here, it's your (as IoT team) module after all.

rlubos commented on lib/at_cmd_parser/at_cmd_parser.c:
```diff
@@ -139,38 +133,46 @@ static int at_parse_process_element(const char **str,
+		if ((*tmpstr == AT_CMD_SEPARATOR) &&
+		    (*(tmpstr + 1) == AT_CMD_READ_TEST_IDENTIFIER)) {
+			tmpstr += 2;
+			list->params[index].type = AT_PARAM_TYPE_CMD_TEST;
```
I'll add a setter/getter for the command type after we agree on how to proceed.

junqingzou commented:
@rlubos API at_params_string_get(), is it possible to add a #define MAX_LENGTH in the header file? It makes the caller easier to reserve buffer.

rlubos commented:
@junqingzou The `at_cmd_parser` uses `malloc` internally to manage memory allocation for strings, so on the parser side the only limitation is a heap size. As for the application using it, it can manage its buffers any way it likes, so it can have a predefined constant specifying buffer length which will be used across the applications. If the string does not fit into the limit, the `at_params_string_get()` call will simply fail.

Another thing is, that by looking at the module I don't see why do we need to copy the string in the first place. The string within `struct at_param_list` will be valid until the `struct at_param_list` is cleared or another command is parsed, therefore the function could simply return a pointer to that string. If the application needs to store it for later, it could do it manually. But that's probably a topic for a different PR.

evenl commented on include/at_cmd_parser/at_params.h:
```diff
@@ -43,6 +43,12 @@ enum at_param_type {
 	/** Empty or optional parameter that should be skipped. */
 	AT_PARAM_TYPE_EMPTY,
+	/** Parameter is a SET AT command. */
+	AT_PARAM_TYPE_CMD_SET,
```
I think it is a bit strange that the at params module should know anything about what command type. I think testing for which command type it is should be a separate helper function.

`
if (at_params_parse_cmd_type(at_params_get_cmd(list)) == AT_CMD_TYPE_TEST) {
        /* Some code */
}
`

rlubos replied:
I've just updated the PR, new version more or less matches what you say, please check.

evenl commented on include/at_cmd_parser/at_params.h:
```diff
@@ -57,8 +72,8 @@ union at_param_value {
 struct at_param {
-	enum at_param_type   type;
-	size_t               size;
+	enum at_param_type type;
```
I still think this is a bit strange, I think the parser should be extended to allow for =,? characters in the command state so the string element would include those characters. Then helper function could figure out if it's an notification, set, read or test command. This would leave the at_parms code unchanged.

rlubos replied:
I disagree - the information about the command type is resolved during the parsing process and should be returned by the parsing function. Hiding this information within the command string will result in:
1) a need to parse this information again,
2) and will make any string operations on the command string more difficult than they already are (like `strcmp`)

evenl replied:
Ok, then we have to agree to disagree:

1. No, you don't need parse the information again, you just accept more information into the string element when parsing a command. It's the same as you need to keep the %, # and + sign to see if it's an proprietary, custom or standard AT command. So what you do is to postpone the parsing of what it is. It simplifies your patch, doesn't change at_params module, keeps the parser code simpler and doesn't open up to add what ever special case into the core parser code.
2. No, because the command is always index 0, so you only need to treat this element specially.

rlubos replied:
> It simplifies your patch, doesn't change at_params module, keeps the parser code simpler and doesn't open up to add what ever special case into the core parser code.

I think we should simplify the API usage, not the patch. I'm not sure what special case you refer to, but different command types covered in this PR are specified by Nordic AT command specification, I just follow the standard here.

> No, because the command is always index 0, so you only need to treat this element specially.

There's no difference if the command string is first on n-th parameter. Having to deal with special characters manually complicates the application code. Let's consider the standard AT+CFUN command as a sample. In case we append special character into the command, in or der to determine what action to take I need to either:
1) Provide 3 copies of the command string, `"AT+CFUN="`, `"AT+CFUN?"`, `"AT+CFUN=?"` and `strcpy` against each of them (for now let's ignore the fact that strings returned by the `at_cmd_parser` are not null-terminated),
2) or `strcpy` against `"AT+CFUN"` and then manually analyze following characters (keeping in mind that some SET commands are empty and have neither of these special characters). I call this parsing.

An alternative is to compare against  `"AT+CFUN"` string and then check the command type field, as proposed in this PR.

And @metc @evenl you should settle in which direction do you want this PR to go. On one hand you request a way more intrusive API change, than current proposal (@metc), on the other hand you don't want me to touch at_cmd_parser code at all (@evenl). Perhaps we should just give up on this patch, and have application to deal manually with AT commands. I don't see what's the point of having the `at_cmd_parser` module then.

evenl replied:
I think you are missunderstanding what I want:

> I think we should simplify the API usage, not the patch.

I think we get both from my proposal. The only change I ask for in your PR is that you move the command type parser (lines 137 through 149) out of the core parser and into a separate function, which is still a part of the at_cmd_parser API. For many users it is not necessary to know what command type it is as they are only parsing notifications, so this sound for me like extra information that doesn't belong in the at_params module context.

> There's no difference if the command string is first on n-th parameter.

Yes, for sure, but the parser is written in a way that will only detect commands at the first index, because that is how AT is specified. In cases with multiple commands, the string will be spitted up in multiple strings where the command always is index 0. 

> Provide 3 copies of the command string, "AT+CFUN=", "AT+CFUN?", "AT+CFUN=?" and strcpy against each of them (for now let's ignore the fact that strings returned by the at_cmd_parser are not null-terminated), or strcpy against "AT+CFUN" and then manually analyze following characters (keeping in mind that some SET commands are empty and have neither of these special characters). I call this parsing.

I think you are complicating the problem, the null-terminator is a decision that was made earlier and is the current implementation of the driver. But anyway in your example, strcmp could have been replaced by
strncmp("AT_CFUN", param, strlen("AT+CFUN");
of
res = strncmp("AT+CFUN", param, MIN(strlen("AT+CFUN"), parm_length); which will work both with and without NULL termination. 

> Having to deal with special characters manually complicates the application code.

That is not what I am trying to convey, I only say that I think the command type parsing should be a separate helper function in the at_cmd_parser API. 

> the other hand you don't want me to touch at_cmd_parser code at all (@evenl). 

I don't say that you should not touch the at_cmd_parser, I only say that you can avoid touching the at_params module that the at_cmd_parser depends on.

But me and Christopher will have a call today, and then we will get back to you.

evenl replied:
Me an Christopher just had a call, and we agreed that we want a separate function to parse the command type. So looking at the PR now, it seems like the only update you need to do is to move the `enum at_cmd_type {}` out of at_params.h and into at_cmd_parser.h and remove `enum at_param_type type;` from the at_parm struct.

rlubos replied:
I've moved `enum at_cmd_type` into the `at_cmd_parser.h` header.

I think you are confused about `enum at_param_type type;` though - this patch has nothing to do with this field and I doubt we want to remove it.

evenl replied:
Aha, you are absolutely right. I was probably just confused because of the alignment change so it came up as a + line.

rlubos commented:
I've updated the PR:
* Removed any command type references from at_param_list,
* Introduced a separate helper function to identify command type based on RAW command string
* Minimal changes in the parser, just to understand the READ/TEST commands.

junqingzou commented on lib/at_cmd_parser/at_cmd_parser.c:
```diff
@@ -343,3 +329,29 @@ int at_parser_max_params_from_str(const char *at_params_str,
+		return AT_CMD_TYPE_UNKNOWN;
+	}
+
+	at_cmd += sizeof("AT+") - 1;
```
There is one AT command, "AT". Will this be OK?

rlubos replied:
Right, I guess we might need some special logic to cover it, I'll have a look.

junqingzou replied:
Move two bytes, then call is_lfcr() to confirm the termination of AT string. If not move one further.

A similar problem in static inline bool is_command(const char *str).

junqingzou commented on lib/at_cmd_parser/at_cmd_parser.c:
```diff
@@ -343,3 +329,29 @@ int at_parser_max_params_from_str(const char *at_params_str,
+	} else if (*at_cmd == AT_CMD_READ_TEST_IDENTIFIER) {
+		type = AT_CMD_TYPE_READ_COMMAND;
+	} else {
+		type = AT_CMD_TYPE_SET_COMMAND;
```
In case of SET command without any parameter, is "at_cmd" pointer moved to the end of the input string? In that case, can we still use * operator?

rlubos replied:
The pointer is a local to this function, any modifications to it are not visible outside.

junqingzou replied:
Again need to call is_lfcr() to confirm the termination of AT string, before using * to access at_cmd pointer.

evenl commented:
Thanks, approved from my side.

rlubos commented:
@junqingzou Pushed a fix for the "lone AT command" case, please check.

junqingzou commented on lib/at_cmd_parser/at_utils.h:
```diff
@@ -49,33 +50,6 @@ static inline bool is_notification(char chr)
- * @brief Check if a string is a beginning of an AT command
- *
- * This function will check if the character is a "AT+" or "AT%" which
- * identifies an AT command.
```
"AT#" as well :)

rlubos replied:
I've already changed this description to `This function will check if the string is a valid AT command prefix.`

junqingzou commented on lib/at_cmd_parser/at_cmd_parser.c:
```diff
@@ -343,3 +340,29 @@ int at_parser_max_params_from_str(const char *at_params_str,
+	while (is_valid_notification_char(*at_cmd)) {
+		at_cmd++;
+	}
+
```
I think there is a need to check CRLF or Terminator here. If yes, then SET type command without parameter.

junqingzou commented on lib/at_cmd_parser/at_cmd_parser.c:
```diff
@@ -343,3 +340,29 @@ int at_parser_max_params_from_str(const char *at_params_str,
+	} else if (*at_cmd == AT_CMD_READ_TEST_IDENTIFIER) {
+		type = AT_CMD_TYPE_READ_COMMAND;
+	} else {
+		type = AT_CMD_TYPE_SET_COMMAND;
```
This "else" could lead to possible error. For example AT+COPS>0 instead of AT+COPS=0, i.e. any character not  AT_CMD_SEPARATOR.

> if (*at_cmd == AT_CMD_SEPARATOR) {
> 	   if  (*(at_cmd + 1) == AT_CMD_READ_TEST_IDENTIFIER) {
> 		type = AT_CMD_TYPE_TEST_COMMAND;
>               } else {
> 		type = AT_CMD_TYPE_SET_COMMAND;
>              }
> 	} else if (*at_cmd == AT_CMD_READ_TEST_IDENTIFIER) {
> 		type = AT_CMD_TYPE_READ_COMMAND;
> 	} else {
> 		type = AT_CMD_TYPE_UNKNOWN;
>     }

rlubos replied:
Ok, I'll rework the `else` condition to be more specific here.

rlubos replied:
Updated, I've kept single-level indentation but added a specific check for `=`/LF/CR/NULL. Otherwise set command type to unknown.
